llama_index/llm_predictor/vellum/prompt_registry.py

```diff
@@ -0,0 +1,208 @@
+
+        If the LlamaIndex prompt hasn't yet been registered, it'll be registered
+        automatically, after which point Vellum becomes the source-of-truth for the
+        prompt's definition.
+
+        In this way, the LlamaIndex prompt is treated as the initial value for the newly
+        registered prompt in Vellum.
+        """
+
+        deployment_id = initial_prompt.prompt_kwargs.get("vellum_deployment_id")
+        deployment_name = initial_prompt.prompt_kwargs.get("vellum_deployment_name")
+
+        registered_prompt: VellumRegisteredPrompt
+
+        if deployment_id or deployment_name:
+            deployment_id_or_name: str = (
+                deployment_id or deployment_name  # type: ignore
+            )
+            registered_prompt = self._get_registered_prompt(deployment_id_or_name)
+        else:
+            registered_prompt = self._register_prompt(initial_prompt)
+
+        return registered_prompt
+
+    def get_compiled_prompt(
+        self, registered_prompt: VellumRegisteredPrompt, input_values: dict[str, Any]
+    ) -> VellumCompiledPrompt:
+        """Retrieves the fully-compiled prompt from Vellum, after all variable
+        substitutions, templating, etc.
+        """
+
+        result = self._vellum_client.model_versions.model_version_compile_prompt(
+            registered_prompt.model_version_id, input_values=input_values
+        )
+        return VellumCompiledPrompt(
+            text=result.prompt.text, num_tokens=result.prompt.num_tokens
+        )
+
+    def _get_registered_prompt(
+        self, deployment_id_or_name: str
+    ) -> VellumRegisteredPrompt:
+        """Retrieves a prompt from Vellum, keying off of the deployment's id/name."""
+
+        deployment = self._vellum_client.deployments.retrieve(deployment_id_or_name)
+
+        # Assume that the deployment backing a registered prompt will always have a
+        # single model version. Note that this may not be true in the future once
+        # deployment-level A/B testing is supported and someone configures an A/B test.
+        model_version_id = deployment.active_model_version_ids[0]
+        model_version = self._vellum_client.model_versions.retrieve(model_version_id)
+
+        sandbox_snapshot_info = model_version.build_config.sandbox_snapshot
+        sandbox_snapshot_id = (
+            sandbox_snapshot_info.id if sandbox_snapshot_info else None
+        )
+        prompt_id = sandbox_snapshot_info.prompt_id if sandbox_snapshot_info else None
+        sandbox_id = sandbox_snapshot_info.sandbox_id if sandbox_snapshot_info else None
+
+        return VellumRegisteredPrompt(
+            deployment_id=deployment.id,
+            deployment_name=deployment.name,
+            model_version_id=model_version.id,
+            sandbox_id=sandbox_id,
+            sandbox_snapshot_id=sandbox_snapshot_id,
+            prompt_id=prompt_id,
+        )
+
+    def _register_prompt(self, prompt: Prompt) -> VellumRegisteredPrompt:
```

Comment:
TODO: Still need to implement the Vellum API that powers the internals of this function.